### PR TITLE
feat: Player camera and movement changed

### PR DESCRIPTION
**Pull request by OriolSors**
- Player camera based on proportional controller
- Sample points for manage camera behaviour
- Player mobility adapted to camera space
- Unattached camera in Player hierarchy

For playing an scene:
- Move the "Basic Camera" to root hierarchy, and tag it as "MainCamera"
- Adjust the xOffset, yOffset, zOffset of the CameraControllerScript as you want. These values are WORLD coordinates and will be the general offset applied to the player
- Create new GameObject in the root hierarchy and attach it to SamplePoints variable in CameraControllerScript
- Put inside this object all the sample points, each one with the script "CameraSample"
- Save scene and load it again

### CODE REVIEW

**Review by PaablooCH — CHANGES_REQUESTED** [review]
i put the playerManagerScript and load the 1.1 but can't execute the game
Imagen de WhatsApp 2023-06-24 a las 00 05 03

**Comment by OriolSors** [comment]
@PaablooCH put the PlayerManager (with 300 as speed value), save scene and load it again. Then you can play the level

**Comment by OriolSors** [comment]
> @PaablooCH put the PlayerManager (with 300 as speed value), save scene and load it again. Then you can play the level

I updated it for make the life easier. Also, now there is only one sample point in the scene for testing reasons. When this will be agreed, approved and adapted for gameplay, I will start to add more.

**Comment by PaablooCH** [comment]
okay, thx

**Comment by PaablooCH** [comment]
this movement happens to you? i move in diagonal 

https://github.com/Horizons-Games/Axolotl-Engine/assets/62140600/2d7da6aa-2477-4fd9-9acb-42d4d0af6e85

**Comment by OriolSors** [comment]
> this movement happens to you? i move in diagonal
> 
>  camara.mp4 
> the camera sample works fine :)

Can you try it in Release?

**Comment by PaablooCH** [comment]
Release engine or Game?

**Comment by OriolSors** [comment]
> Release engine or Game?

ReleaseEngine

**Comment by markmeliaa** [comment]
So this only works on release now? Any idea why it's so wacky in debug?

**Comment by OriolSors** [comment]
> So this only works on release now? Any idea why it's so wacky in debug?

It's because of framerate. The controller of the camera focus the player position instead of being "dragged" by him. So if the position of the player changes drastically between frames, the camera will get crazy.

**Comment by markmeliaa** [comment]
Oh, understandable! Although this may hinder our future trials in debug due to the camera being slow, that's what we were asked to do, so I guess its fine (we can always improve the framerate in debug to fix that too). Let me check it's working correctly and I'll approve this pr real quick!

**Review by markmeliaa — COMMENTED** [review]
In level 1.1 everything works correctly, but in the gameplay scene the engine breaks when I try to move Bix after pressing play, is something missing in that scene? What should we add to make it work correctly?

**Comment by OriolSors** [comment]
> In level 1.1 everything works correctly, but in the gameplay scene the engine breaks when I try to move Bix after pressing play, is something missing in that scene? What should we add to make it work correctly?

I put the basic init conditions for the scene! Try it now! I will update the description of the PR with the steps needed, sorry!

**Comment by OriolSors** [comment]
> Oh, understandable! Although this may hinder our future trials in debug due to the camera being slow, that's what we were asked to do, so I guess its fine (we can always improve the framerate in debug to fix that too). Let me check it's working correctly and I'll approve this pr real quick!

Yess, in fact with few assets, and a very simple scene we will be able to test it correctly. The problem is when we get 4 FPS, that won't allow us to test anything, not only by the camera

**Review by DADSILVER — APPROVED** [review]
Awesome job, only I think When I try to pass the park out the camera don let us see where is the next jump. I don't know if you can fix it or is something that you have to discuss with level.

I have these 2 questions too.

**Comment by OriolSors** [comment]
> Awesome job, only I think When I try to pass the park out the camera don let us see where is the next jump. I don't know if you can fix it or is something that you have to discuss with level.
> 
> I have these 2 questions too.

Well, actually there are some sample points missing. I added a few of them only for testing, now that all will be approved I will start adding more, and then Level Team will bring them to the new version of the scene

File: Game/Scripts/PlayerMoveScript.cpp [hunk 1]
```diff
@@ -113,12 +116,21 @@ void PlayerMoveScript::Move(float deltaTime)
...
 		totalDirection = totalDirection.Normalized();
+		
+		btTransform worldTransform = btRb->getWorldTransform();
+		Quat rot = Quat::LookAt(componentTransform->GetGlobalForward().Normalized(), totalDirection, float3::unitY, float3::unitY);
+		rot = rot * componentTransform->GetGlobalRotation();
+		worldTransform.setRotation({ rot.x, rot.y, rot.z, rot.w });
+		btRb->setWorldTransform(worldTransform);
+		btRb->getMotionState()->setWorldTransform(worldTransform);
+
 		movement = btVector3(totalDirection.x, totalDirection.y, totalDirection.z) * deltaTime * newSpeed;
```
[the remark this answers is not in the document]
Comment (OriolSors) [reply Game/Scripts/PlayerMoveScript.cpp:134]: Actually I'm currently working on that (some troubles appear). This function takes de totalDirection as the final target for the transform global forward, but as we are using only discrete directions with WASD, it seems not as soft as if we use a joystick. For avoiding that, we will need to use, maybe the SetPositionTarget (and I tried and it works, but sometimes depending of framerate, the collisions are not managed well; similar to the force problem), or interpolating the rotation with a proportional controller for the transform rotation. This is the approach I'm following but there are few difficulties

File: Game/Scripts/CameraSample.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+{
+	REGISTER_FIELD(influenceRadius, float);
+	REGISTER_FIELD(positionOffset, float3);
+	REGISTER_FIELD(orientationOffset, float3);
+}
+
+void CameraSample::Start()
+{
+	transform = owner->GetComponent<ComponentTransform>();
+	position = transform->GetGlobalPosition();
```
[the remark this answers is not in the document]
Comment (OriolSors) [reply Game/Scripts/CameraSample.cpp:26]: Yess, makes sense. In the start there were a lot of different parameters while trying to adjust the sample structure, and finally I left as this. I will make a rework on that

**Review by PaablooCH — APPROVED** [review]

**Review by markmeliaa — APPROVED** [review]
Everything perfect Oriol, amazing work!

**Pull request merged**